File: android_frontend/app/src/main/java/com/tracknrace/android/NativePairDiscovery.java
package com.tracknrace.android;

final class NativePairDiscovery {
    interface Listener {
        void onService(Service service);
    }

    static final class Service {
        final String serverId;
        final String name;
        final String address;
        final int port;
        final boolean pairing;

        Service(String serverId, String name, String address, int port, boolean pairing) {
            this.serverId = serverId;
            this.name = name;
            this.address = address;
            this.port = port;
            this.pairing = pairing;
        }
    }

    static { System.loadLibrary("tracknrace_android"); }

    private final Listener listener;

    NativePairDiscovery(Listener listener) { this.listener = listener; }
    String start() { return nativeStart(); }
    void stop() { nativeStop(); }

    @SuppressWarnings("unused")
    private void onNativeService(String serverId, String name, String address,
                                 int port, boolean pairing) {
        listener.onService(new Service(serverId, name, address, port, pairing));
    }

    private native String nativeStart();
    private native void nativeStop();
}

File: android_frontend/app/src/main/java/com/tracknrace/android/NativePairDiscovery.kt
package com.tracknrace.android

internal class NativePairDiscovery(
    private val listener: Listener,
) {
    interface Listener {
        fun onService(service: Service)
    }

    data class Service(
        val serverId: String,
        val name: String,
        val address: String,
        val port: Int,
        val pairing: Boolean,
    )

    companion object {
        init {
            System.loadLibrary("tracknrace_android")
        }
    }

    fun start(): String? = nativeStart()

    fun stop() = nativeStop()

    @Suppress("unused") // Invoked by JNI on the discovery worker thread.
    private fun onNativeService(
        serverId: String,
        name: String,
        address: String,
        port: Int,
        pairing: Boolean,
    ) {
        listener.onService(Service(serverId, name, address, port, pairing))
    }

    private external fun nativeStart(): String?

    private external fun nativeStop()
}
